include/mesh.tcc

```diff
@@ -299,22 +298,104 @@ void mpm::Mesh<Tdim>::compute_cell_neighbours() {
   }
 
   // Assign neighbour to cells
-  tbb::parallel_for(
-      tbb::blocked_range<int>(size_t(0), size_t(cells_.size()),
-                              tbb_grain_size_),
-      [&](const tbb::blocked_range<int>& range) {
-        for (int i = range.begin(); i != range.end(); ++i) {
-          // Iterate over each node in current cell
-          for (auto id : cells_[i]->nodes_id()) {
-            auto cell_id = cells_[i]->id();
-            // Get the cells associated with each node
-            for (auto neighbour_id : node_cell_map[id])
-              if (neighbour_id != cell_id)
-                cells_[i]->add_neighbour(neighbour_id);
-          }
-        }
-      },
-      tbb::simple_partitioner());
+  tbb::parallel_for(tbb::blocked_range<int>(size_t(0), size_t(cells_.size()),
+                                            tbb_grain_size_),
+                    [&](const tbb::blocked_range<int>& range) {
+                      for (int i = range.begin(); i != range.end(); ++i) {
+                        // Iterate over each node in current cell
+                        for (auto id : cells_[i]->nodes_id()) {
+                          auto cell_id = cells_[i]->id();
+                          // Get the cells associated with each node
+                          for (auto neighbour_id : node_cell_map[id])
+                            if (neighbour_id != cell_id)
+                              cells_[i]->add_neighbour(neighbour_id);
+                        }
+                      }
+                    },
+                    tbb::simple_partitioner());
+}
+
+//! Find particle neighbours for all particle
+template <unsigned Tdim>
+void mpm::Mesh<Tdim>::compute_particle_neighbours() {
+  // Check whether cells have been assigned with neighbours
+  if (!(*cells_.cbegin())->has_neighbours()) this->compute_cell_neighbours();
+
+  // Loop over cells
+  for (auto citr = cells_.cbegin(); citr != cells_.cend(); ++citr) {
+    // Loop over the neighbouring cell particles
+    std::vector<mpm::Index> neighbouring_particle_sets = (*citr)->particles();
+    for (const auto& neighbour_cell_id : (*citr)->neighbours())
+      neighbouring_particle_sets.insert(
+          neighbouring_particle_sets.end(),
+          map_cells_[neighbour_cell_id]->particles().begin(),
+          map_cells_[neighbour_cell_id]->particles().end());
+
+    // Loop over the particles in the current cells and assign particle
+    // neighbours
+    for (const auto& p_id : (*citr)->particles()) {
+      bool status =
+          map_particles_[p_id]->assign_neighbours(neighbouring_particle_sets);
+      if (!status)
+        throw std::runtime_error("Cannot assign valid particle neighbours");
+    }
+  }
+}
+
+//! Find particle neighbours for all particle in a given cell
+template <unsigned Tdim>
+void mpm::Mesh<Tdim>::compute_particle_neighbours(const Cell<Tdim>& cell) {
```

Comment:
use a default function argument as a nullptr, so if nothing is passed it would be the same function as above.

Reply:
without default `nullptr`, wouldn't it work too?

Reply:
I made default `nullptr` in the header. Resolved

Reply:
I was thinnking of merging this function and the one above, they are very similar

Reply:
```
mpm::Container<mpm::Cell> cell_ctr;
cell_ctr.add(cell);
auto cells = (cell == nullptr) ? cells_ : cell_ctr;
```